File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for i in range(len(asteroids)):
            if stack == [] or stack[-1] < 0:
                stack.append(asteroids[i])
            else:
                while True:
                    if stack[-1] * asteroids[i] > 0:
                        stack.append(asteroids[i])
                        break
                    elif stack[-1] + asteroids[i] > 0:
                        break
                    elif stack[-1] + asteroids[i] == 0:
                        stack.pop(-1)
                        break
                    elif stack[-1] + asteroids[i] < 0:
                        stack.pop(-1)
                        if stack == []:
                            stack.append(asteroids[i])
                            break
        
        return stack
    

File: 0735-asteroid-collision/0735-asteroid-collision.rb
# @param {Integer[]} asteroids
# @return {Integer[]}
def asteroid_collision(asteroids)
    stack = []
    (0...asteroids.length).each do |i|
        if stack == [] || stack[-1] < 0
            stack << asteroids[i]
        else
            while true
                if stack[-1] * asteroids[i] > 0
                    stack << asteroids[i]
                    break
                elsif stack[-1] + asteroids[i] > 0
                    break
                elsif stack[-1] + asteroids[i] == 0
                    stack.pop()
                    break
                elsif stack[-1] + asteroids[i] < 0
                    stack.pop()
                    if stack == []
                        stack << asteroids[i]
                        break
                    end
                end
            end
        end
    end
    return stack
end